File: internal/checks_lua/oauth_discovery.lua
-- oauth-discovery: Lua port of internal/checks/oauth_discovery.go.
--
-- Passive audit of the OAuth 2.0 Authorization Server Metadata
-- (RFC 8414) and OpenID Connect Discovery 1.0 documents an identity
-- provider publishes at well-known paths on its issuer host. The
-- documents declare which signing algorithms, client-auth methods,
-- PKCE methods, and response types the AS accepts; misconfigurations
-- in these advertised values produce real attacker primitives even
-- before any login flow is exercised.
--
-- One finding per detected weakness (alg=none, symmetric algs, only
-- token-auth=none, missing/plain-only PKCE, implicit-flow response
-- types, plain-HTTP endpoints). Per-host caching lives in Go on
-- ctx.oauth.discover so a 50-page crawl probes the well-known
-- endpoint once; the audit policy runs in Lua against the cached
-- facts and the finding catalog is composed here, so the rule's
-- prose / severity policy is editable without rebuilding Go.

local check = {
  name  = "oauth-discovery",
  level = "passive",
  scope = "host",
}

-- BODY_SNIPPET_CAP matches Go's snippetJSON cap (512 bytes) so the
-- finding evidence reads identically across implementations. The full
-- body lives in the Exchange field; this is only the inline snippet.
local BODY_SNIPPET_CAP = 512

local function snippet_json(body)
  if body == nil or body == "" then return "" end
  -- TrimSpace mirrors bytes.TrimSpace in the Go path so leading /
  -- trailing whitespace from the JSON envelope does not appear in
  -- the snippet.
  local trimmed = string.gsub(body, "^%s+", "")
  trimmed = string.gsub(trimmed, "%s+$", "")
  if #trimmed > BODY_SNIPPET_CAP then
    return string.sub(trimmed, 1, BODY_SNIPPET_CAP)
  end
  return trimmed
end

local function lower_set(arr)
  local out = {}
  if arr == nil then return out end
  for _, s in ipairs(arr) do
    local trimmed = string.gsub(s, "^%s+", "")
    trimmed = string.gsub(trimmed, "%s+$", "")
    out[string.lower(trimmed)] = true
  end
  return out
end

local function only_contains(set, want)
  local n = 0
  for k in pairs(set) do
    n = n + 1
    if k ~= want then return false end
  end
  return n > 0
end

local function symmetric_algs(algs)
  local out = {}
  for _, candidate in ipairs({"hs256", "hs384", "hs512"}) do
    if algs[candidate] then
      out[#out + 1] = string.upper(candidate)
    end
  end
  return out
end

local function pkce_weakness(methods)
  local n = 0
  for _ in pairs(methods) do n = n + 1 end
  if n == 0 then
    return "does not advertise code_challenge_methods_supported, so PKCE is not announced as a capability"
  end
  if methods["s256"] then return "" end
  if methods["plain"] then
    return 'advertises only "plain" in code_challenge_methods_supported, which provides no protection against a code interceptor'
  end
  return "advertises code_challenge_methods_supported without S256"
end

local function implicit_flow_types(types)
  local out = {}
  -- "code id_token" is a hybrid that includes the code path and is
  -- not strictly implicit, but the fragment leak applies whenever
  -- id_token rides in the URL response, so it gets flagged too. Mirrors
  -- the Go ordering verbatim so finding details render identically.
  for _, rt in ipairs({"token", "id_token", "id_token token", "token id_token"}) do
    if types[rt] then
      out[#out + 1] = rt
    end
  end
  return out
end

local function plain_http_endpoints(facts)
  local out = {}
  local function check_field(label, raw)
    if raw == nil or raw == "" then return end
    local u = ctx_url_parse_safe(raw)
    if u and string.lower(u.scheme) == "http" then
      out[#out + 1] = label .. "=" .. raw
    end
  end
  check_field("authorization_endpoint", facts.authorization_endpoint)
  check_field("token_endpoint",         facts.token_endpoint)
  check_field("userinfo_endpoint",      facts.userinfo_endpoint)
  check_field("jwks_uri",               facts.jwks_uri)
  check_field("introspection_endpoint", facts.introspection_endpoint)
  check_field("revocation_endpoint",    facts.revocation_endpoint)
  return out
end

-- ctx_url_parse_safe is filled in once we hold the ctx (we need its
-- helper table). Defined as a forward-declared local because the
-- plain-HTTP check above closes over it without an explicit ctx
-- parameter.
ctx_url_parse_safe = nil

-- format_string_array mirrors Go's `fmt.Sprintf("%v", []string{...})`
-- which produces "[a b c]" for the finding detail prose. The Go
-- check folds the advertised values straight into the Sprintf with
-- %v; we replicate that bracketed-space-joined form so the Lua
-- finding text byte-matches the Go one.
local function format_string_array(arr)
  if arr == nil or #arr == 0 then return "[]" end
  return "[" .. table.concat(arr, " ") .. "]"
end

local function build_evidence(ctx, facts)
  return ctx.evidence.build {
    method  = "GET",
    url     = facts.probe_url,
    status  = facts.status,
    body    = snippet_json(facts.body),
  }
end

local function finding_alg_none(ctx, facts)
  return {
    severity = ctx.severity.critical,
    target   = facts.probe_url,
    url      = facts.probe_url,
    title    = "OAuth/OIDC discovery advertises alg=none for id_token",
    detail   = string.format(
      'The authorization server at %s lists "none" in id_token_signing_alg_values_supported '
        .. '(values: %s). An RP that pins acceptable algorithms against the advertised set will accept '
        .. 'unsigned id_tokens, letting any caller forge claims by sending an alg=none token with an empty '
        .. 'signature. The vulnerability lives in every RP that trusts this AS, not just one client.',
      facts.issuer, format_string_array(facts.id_token_signing_alg_values_supported)),
    cwe         = "CWE-327",
    owasp       = "A02:2021 Cryptographic Failures",
    remediation = 'Remove "none" from id_token_signing_alg_values_supported. There is no production use case '
      .. 'for unsigned id_tokens; an unsigned token provides no integrity guarantee and an attacker can substitute '
      .. 'arbitrary claims. Configure the AS to advertise only asymmetric algorithms (RS256, ES256, EdDSA) and '
      .. 'reissue rotated keys via jwks_uri.',
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = { "alg-none" },
    },
  }
end

local function finding_symmetric_alg(ctx, facts, sym_algs)
  return {
    severity = ctx.severity.medium,
    target   = facts.probe_url,
    url      = facts.probe_url,
    title    = "OAuth/OIDC discovery advertises symmetric id_token signing",
    detail   = string.format(
      'The authorization server at %s advertises symmetric id_token signing algorithms (%s) in '
        .. 'id_token_signing_alg_values_supported. Symmetric algorithms require the AS and every relying party '
        .. 'to share the same secret to verify tokens, so one RP\'s secret compromise lets that RP forge tokens '
        .. 'any other RP will accept. The OIDC core spec deprecated HS* outside narrow same-trust-domain '
        .. 'deployments for this reason.',
      facts.issuer, format_string_array(sym_algs)),
    cwe         = "CWE-327",
    owasp       = "A02:2021 Cryptographic Failures",
    remediation = 'Migrate to asymmetric id_token signing (RS256, ES256, EdDSA). Publish the public key via '
      .. 'jwks_uri so RPs verify against a key only the AS holds the private half of. If a symmetric algorithm '
      .. 'must remain for a legacy client, advertise it only for that client\'s audience rather than as a server-'
      .. 'wide capability.',
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = { "symmetric-alg" },
    },
  }
end

local function finding_token_endpoint_auth_none(ctx, facts)
  return {
    severity = ctx.severity.high,
    target   = facts.probe_url,
    url      = facts.probe_url,
    title    = "OAuth/OIDC token endpoint accepts only unauthenticated clients",
    detail   = string.format(
      'The authorization server at %s advertises only "none" in token_endpoint_auth_methods_supported. '
        .. 'That means the token endpoint will mint tokens for any caller presenting a valid authorization code '
        .. 'without verifying client identity, so an attacker who intercepts a code can trade it for tokens '
        .. 'indistinguishably from the legitimate client. Confidential clients become impossible against this AS.',
      facts.issuer),
    cwe         = "CWE-287",
    owasp       = "A07:2021 Identification and Authentication Failures",
    remediation = 'Configure the AS to support a real client-auth method for confidential clients '
      .. '(client_secret_basic, client_secret_post, private_key_jwt). Reserve token_endpoint_auth_method=none '
      .. 'for public clients (SPAs, native apps) that pair it with PKCE; even then, confidential clients should '
      .. 'have a stronger option available.',
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = { "token-auth-none" },
    },
  }
end

local function finding_pkce_weak(ctx, facts, weakness)
  return {
    severity = ctx.severity.medium,
    target   = facts.probe_url,
    url      = facts.probe_url,
    title    = "OAuth/OIDC discovery advertises weak or absent PKCE support",
    detail   = string.format(
      'The authorization server at %s %s. PKCE binds an authorization code to the client that requested it, '
        .. 'preventing an interceptor of the code from trading it for tokens. Without S256 enforcement, public '
        .. 'clients (SPAs, native apps) fall back to bearer-style code exchange and any party who reads the '
        .. 'redirect URL can complete the flow.',
      facts.issuer, weakness),
    cwe         = "CWE-287",
    owasp       = "A07:2021 Identification and Authentication Failures",
    remediation = 'Advertise S256 in code_challenge_methods_supported and reject authorization requests without '
      .. 'a code_challenge parameter for public clients. OAuth 2.1 and FAPI 2.0 mandate PKCE with S256; legacy '
      .. '"plain" support should be removed since it provides no protection against a code interceptor.',
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = { "pkce-weak" },
    },
  }
end

local function finding_implicit_flow(ctx, facts, types)
  return {
    severity = ctx.severity.low,
    target   = facts.probe_url,
    url      = facts.probe_url,
    title    = "OAuth/OIDC discovery advertises deprecated implicit flow",
    detail   = string.format(
      'The authorization server at %s advertises implicit-flow response types (%s) in '
        .. 'response_types_supported. The implicit flow lands access tokens (and sometimes id_tokens) in the '
        .. 'URL fragment, where they leak through browser history, server access logs, the Referer header, and '
        .. 'document.location. OAuth 2.1 and the OIDC "implicit considered harmful" guidance recommend the '
        .. 'authorization code flow with PKCE for every client shape that previously used implicit.',
      facts.issuer, format_string_array(types)),
    cwe         = "CWE-598",
    owasp       = "A04:2021 Insecure Design",
    remediation = 'Stop advertising implicit response types. Migrate SPA / native clients to authorization code '
      .. 'flow with PKCE, which keeps tokens out of the URL and supports refresh tokens. If a client cannot be '
      .. 'migrated immediately, scope the deprecation to that client\'s metadata rather than as a server-wide '
      .. 'capability.',
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = { "implicit-flow" },
    },
  }
end

local function finding_plain_http_endpoint(ctx, facts, endpoints)
  return {
    severity = ctx.severity.high,
    target   = facts.probe_url,
    url      = facts.probe_url,
    title    = "OAuth/OIDC discovery advertises endpoints over plain HTTP",
    detail   = string.format(
      'The authorization server at %s advertises one or more endpoints over plain HTTP (%s). Any caller '
        .. 'on the network between the user agent and the AS can read or rewrite the authorization request, '
        .. 'the code exchange, or the userinfo response. OAuth 2.0 (RFC 6749) and OIDC core both require TLS '
        .. 'on every endpoint in the flow.',
      facts.issuer, format_string_array(endpoints)),
    cwe         = "CWE-319",
    owasp       = "A02:2021 Cryptographic Failures",
    remediation = 'Serve every OAuth / OIDC endpoint over HTTPS and update the discovery document so the '
      .. 'published URLs match. If the AS is behind a TLS-terminating proxy, ensure the metadata advertises the '
      .. 'external HTTPS URL rather than the internal HTTP one.',
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = { "plain-http" },
    },
  }
end

-- restamp_to_page mirrors the Go restampFindings pass: the audit
-- builds each finding against the discovery probe URL (so the dedupe
-- key keys off the canonical resource) but we re-stamp Target / URL
-- to the current page so the report ties to a URL the operator saw.
local function restamp_to_page(findings, page_url)
  for _, f in ipairs(findings) do
    f.target = page_url
    f.url    = page_url
  end
end

function check.run(ctx)
  ctx_url_parse_safe = function(raw)
    local u, _ = ctx.url.parse(raw)
    return u
  end

  local facts, err = ctx.oauth.discover(ctx.page.url)
  if err then
    ctx:report("oauth-discovery: " .. err)
    return nil
  end
  if facts == nil then return nil end

  local findings = {}

  local algs = lower_set(facts.id_token_signing_alg_values_supported)
  if algs["none"] then
    findings[#findings + 1] = finding_alg_none(ctx, facts)
  end
  local sym = symmetric_algs(algs)
  if #sym > 0 then
    findings[#findings + 1] = finding_symmetric_alg(ctx, facts, sym)
  end

  local auth_methods = lower_set(facts.token_endpoint_auth_methods_supported)
  if only_contains(auth_methods, "none") then
    findings[#findings + 1] = finding_token_endpoint_auth_none(ctx, facts)
  end

  local pkce_methods = lower_set(facts.code_challenge_methods_supported)
  local pkce_weak = pkce_weakness(pkce_methods)
  if pkce_weak ~= "" then
    findings[#findings + 1] = finding_pkce_weak(ctx, facts, pkce_weak)
  end

  local resp_types = lower_set(facts.response_types_supported)
  local implicit = implicit_flow_types(resp_types)
  if #implicit > 0 then
    findings[#findings + 1] = finding_implicit_flow(ctx, facts, implicit)
  end

  local plain = plain_http_endpoints(facts)
  if #plain > 0 then
    findings[#findings + 1] = finding_plain_http_endpoint(ctx, facts, plain)
  end

  if #findings == 0 then return nil end
  restamp_to_page(findings, ctx.page.url)
  return findings
end

return check

File: internal/lua_engine/checks/platform/oauth/oauth_discovery.go
package oauth

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strings"
	"sync"

	"github.com/londonmax12/hyperz/internal/httpclient"
	"github.com/londonmax12/hyperz/internal/scope"
)

// OAuthDiscovery audits the public OAuth 2.0 Authorization Server
// Metadata (RFC 8414) and OpenID Connect Discovery 1.0 documents an
// identity provider publishes at well-known paths on its issuer host.
// The documents declare which signing algorithms, client-auth methods,
// PKCE methods, and response types the AS accepts; misconfigurations
// in these advertised values produce real attacker primitives even
// before any login flow is exercised.
//
// What the check looks for (one finding per hit):
//
//  1. id_token_signing_alg_values_supported contains "none" - the AS
//     advertises that it will issue unsigned id_tokens, which a
//     validator pinning alg against the advertised set will accept.
//     Critical.
//  2. id_token_signing_alg_values_supported contains a symmetric
//     algorithm (HS256/HS384/HS512) - the secret is shared with every
//     RP, which is the wrong trust model for a federation. Medium.
//  3. token_endpoint_auth_methods_supported is { "none" } only - the
//     AS will issue tokens to any caller that presents a code, making
//     every public client interchangeable with every other. High.
//  4. code_challenge_methods_supported missing or { "plain" } only -
//     PKCE is not enforced, or only a no-op "transformation" is
//     offered. Medium.
//  5. response_types_supported includes "token" or "id_token" (the
//     implicit flow) - deprecated by RFC 9700 / OAuth 2.1 because
//     tokens land in the URL fragment and leak to history, logs, and
//     referrers. Low.
//  6. Any AS endpoint URL (authorization, token, userinfo, jwks_uri,
//     introspection, revocation) advertised over plain HTTP - the
//     entire flow becomes interceptable. High.
//
// Per-host: the check fetches each well-known path at most once per
// scan, caches the result, and re-emits cached findings against
// subsequent pages on the same host with the new page URL attached so
// the report ties the finding to a page the user actually saw.
//
// Out of scope (require operator-supplied client credentials to test):
//   - state parameter validation
//   - redirect_uri strict matching
//   - end-to-end PKCE enforcement (vs. the advertised method support)
//   - actual id_token signature verification by RP
//
// These would need a real client_id and redirect_uri to drive a
// flow; the check intentionally limits itself to evidence the
// discovery document already publishes.
//
// Passive (LevelPassive) check.
type OAuthDiscovery struct {
	once  sync.Once
	mu    sync.Mutex
	cache map[string]oauthDiscoveryCacheEntry
}

// oauthDiscoveryCacheEntry memoizes the per-host probe result. A nil
// facts pointer represents a confirmed negative (well-known paths
// 404, malformed JSON, no document); a populated facts is re-used
// across every page on the host so the well-known endpoint is hit at
// most once per scan.
type oauthDiscoveryCacheEntry struct {
	facts *OAuthDiscoveryFacts
}

// OAuthDiscoveryFacts is the raw scan-facts shape the bridge returns
// to the Lua port. Mirrors the subset of the discovery document the
// audit policy actually reads plus the probe metadata needed for
// evidence rendering. The audit (alg=none, symmetric algs, PKCE
// weakness, implicit flow, plain-HTTP endpoints) lives in the .lua
// file; this struct is the algorithm output, not a finding shape.
type OAuthDiscoveryFacts struct {
	Issuer                            string
	AuthorizationEndpoint             string
	TokenEndpoint                     string
	UserinfoEndpoint                  string
	JwksURI                           string
	IntrospectionEndpoint             string
	RevocationEndpoint                string
	ResponseTypesSupported            []string
	IDTokenSigningAlgValuesSupported  []string
	TokenEndpointAuthMethodsSupported []string
	CodeChallengeMethodsSupported     []string
	ProbeURL                          string
	Status                            int
	Body                              []byte
}

const (
	// oauthDiscoveryBodyCap bounds the JSON body the check buffers.
	// Real-world OIDC discovery documents (Google, Okta, Auth0, Azure
	// AD) all land well under 16 KiB; 64 KiB clears even the
	// pathological pretty-printed cases without letting a misbehaving
	// edge pin the worker on a slow stream.
	oauthDiscoveryBodyCap = 64 << 10
)

// oauthDiscoveryPaths are the well-known suffixes the canonical
// "oidc" catalogue probes on the issuer host. The two specs share
// a path convention; both are tried because some servers expose one
// and not the other (Okta publishes OIDC discovery only, certain
// plain-OAuth-only deployments publish RFC 8414 only). Kept as a
// package-level slice so the "oidc" catalogue refers to it by
// name rather than re-defining the strings inline.
var oauthDiscoveryPaths = []string{
	"/.well-known/openid-configuration",
	"/.well-known/oauth-authorization-server",
}

// oauthDiscoveryCatalogue is a named bundle of well-known paths an
// OAuth-discovery-shaped check probes per host. The Lua bridge takes
// the name as ctx.oauth.discover's first argument and resolves it
// through resolveOAuthDiscoveryCatalogue. A future check that needs
// to probe a non-standard discovery surface (a vendor-specific
// well-known suffix, an RFC-9728 OAuth-protected-resource document)
// registers its own catalogue here instead of editing the canonical
// probe path list.
type oauthDiscoveryCatalogue struct {
	paths []string
}

// oauthDiscoveryCatalogues is the named-catalogue registry. "oidc"
// covers the canonical RFC 8414 + OIDC discovery 1.0 well-known
// paths; sibling catalogues add themselves to this map and the Lua
// bridge surfaces them automatically.
var oauthDiscoveryCatalogues = map[string]oauthDiscoveryCatalogue{
	"oidc": {paths: oauthDiscoveryPaths},
}

// resolveOAuthDiscoveryCatalogue returns the named catalogue, falling
// back to "oidc" when name is empty or unknown. Same typo-tolerance
// rule resolveDiscoveryCatalogue and resolveSmugglingCatalogue use.
func resolveOAuthDiscoveryCatalogue(name string) oauthDiscoveryCatalogue {
	if cat, ok := oauthDiscoveryCatalogues[name]; ok {
		return cat
	}
	return oauthDiscoveryCatalogues["oidc"]
}

// oauthDiscoveryDoc is the subset of fields the check inspects.
// Unknown fields are ignored; the document carries dozens of values
// but only these have actionable security signal at the metadata level.
type oauthDiscoveryDoc struct {
	Issuer                            string   `json:"issuer"`
	AuthorizationEndpoint             string   `json:"authorization_endpoint"`
	TokenEndpoint                     string   `json:"token_endpoint"`
	UserinfoEndpoint                  string   `json:"userinfo_endpoint"`
	JwksURI                           string   `json:"jwks_uri"`
	IntrospectionEndpoint             string   `json:"introspection_endpoint"`
	RevocationEndpoint                string   `json:"revocation_endpoint"`
	ResponseTypesSupported            []string `json:"response_types_supported"`
	IDTokenSigningAlgValuesSupported  []string `json:"id_token_signing_alg_values_supported"`
	TokenEndpointAuthMethodsSupported []string `json:"token_endpoint_auth_methods_supported"`
	CodeChallengeMethodsSupported     []string `json:"code_challenge_methods_supported"`
}

// DiscoverFacts returns the cached or freshly fetched discovery facts
// for the host implied by pageURL, or nil when no well-known path
// served a parseable document. catalogue selects which registered
// well-known path bundle to probe ("oidc" for the canonical RFC
// 8414 + OIDC discovery surface); unknown names fall back to
// "oidc" via resolveOAuthDiscoveryCatalogue. Per-host cache
// lifetime matches this receiver's lifetime (one *OAuthDiscovery per
// scan registration).
//
// The Lua port reads these facts and composes the finding catalog
// itself (title / severity / detail / CWE / OWASP / remediation /
// dedupe key / evidence); the algorithm input (HTTP fetch + JSON
// parse) stays in Go so per-host work happens at most once per scan
// regardless of which implementation runs at scan time.
func (c *OAuthDiscovery) DiscoverFacts(ctx context.Context, client *httpclient.Client, sc *scope.Scope, pageURL, catalogue string) (*OAuthDiscoveryFacts, error) {
	c.once.Do(func() {
		c.cache = map[string]oauthDiscoveryCacheEntry{}
	})

	u, err := url.Parse(pageURL)
	if err != nil || u.Scheme == "" || u.Host == "" {
		return nil, nil
	}
	if !sc.Allows(u) {
		return nil, nil
	}
	hostKey := strings.ToLower(u.Scheme + "://" + u.Host)

	c.mu.Lock()
	entry, cached := c.cache[hostKey]
	c.mu.Unlock()
	if cached {
		return entry.facts, nil
	}

	facts := c.probeHost(ctx, client, sc, u, resolveOAuthDiscoveryCatalogue(catalogue))
	c.mu.Lock()
	c.cache[hostKey] = oauthDiscoveryCacheEntry{facts: facts}
	c.mu.Unlock()
	return facts, nil
}

// probeHost fetches each well-known path in the resolved catalogue
// against u's host until one returns a parseable discovery document.
// Returns the parsed facts, or nil if no path was reachable /
// parseable.
func (c *OAuthDiscovery) probeHost(ctx context.Context, client *httpclient.Client, sc *scope.Scope, u *url.URL, cat oauthDiscoveryCatalogue) *OAuthDiscoveryFacts {
	base := u.Scheme + "://" + u.Host
	for _, path := range cat.paths {
		probeURL := base + path
		probeU, err := url.Parse(probeURL)
		if err != nil {
			continue
		}
		if !sc.Allows(probeU) {
			continue
		}
		doc, status, body, ok := c.fetchDoc(ctx, client, probeURL)
		if !ok {
			continue
		}
		return &OAuthDiscoveryFacts{
			Issuer:                            doc.Issuer,
			AuthorizationEndpoint:             doc.AuthorizationEndpoint,
			TokenEndpoint:                     doc.TokenEndpoint,
			UserinfoEndpoint:                  doc.UserinfoEndpoint,
			JwksURI:                           doc.JwksURI,
			IntrospectionEndpoint:             doc.IntrospectionEndpoint,
			RevocationEndpoint:                doc.RevocationEndpoint,
			ResponseTypesSupported:            doc.ResponseTypesSupported,
			IDTokenSigningAlgValuesSupported:  doc.IDTokenSigningAlgValuesSupported,
			TokenEndpointAuthMethodsSupported: doc.TokenEndpointAuthMethodsSupported,
			CodeChallengeMethodsSupported:     doc.CodeChallengeMethodsSupported,
			ProbeURL:                          probeURL,
			Status:                            status,
			Body:                              body,
		}
	}
	return nil
}

// fetchDoc GETs probeURL and parses the body as an OIDC / OAuth
// discovery document. Returns ok=false on transport error, non-200,
// non-JSON, or a JSON object that does not carry the issuer field
// (the one MUST in both specs - its presence confirms we hit a real
// metadata document rather than a generic JSON 404).
func (c *OAuthDiscovery) fetchDoc(ctx context.Context, client *httpclient.Client, probeURL string) (*oauthDiscoveryDoc, int, []byte, bool) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, probeURL, nil)
	if err != nil {
		return nil, 0, nil, false
	}
	req.Header.Set("Accept", "application/json")
	resp, err := client.Do(ctx, req)
	if err != nil {
		return nil, 0, nil, false
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		// Drain a small chunk to release the connection cleanly. We
		// don't surface non-200 as an error - both well-known paths
		// are expected to 404 on non-IdP hosts.
		_, _, _ = httpclient.ReadBodyCapped(resp, 1<<10)
		return nil, resp.StatusCode, nil, false
	}
	body, _, err := httpclient.ReadBodyCapped(resp, oauthDiscoveryBodyCap)
	if err != nil {
		return nil, resp.StatusCode, nil, false
	}
	var doc oauthDiscoveryDoc
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, resp.StatusCode, nil, false
	}
	if doc.Issuer == "" {
		// The issuer claim is REQUIRED in both RFC 8414 and OIDC
		// Discovery. A "document" without it is almost certainly a
		// generic JSON 404 / WAF interstitial; treat it as a miss
		// rather than emit findings against fabricated nil-array
		// values.
		return nil, resp.StatusCode, nil, false
	}
	return &doc, resp.StatusCode, body, true
}
